fix(ivy): pass ngContentSelectors through to `defineComponent()` calls

petebacondarwin opened the pull request:
Scenarios like ngUpgrade need to pass blocks of projected content
through to components that they dynamically instantiate.

The instantiator of the component needs to know what CSS selectors
to use in order to select the correct blocks of content to be projected.

This change takes the list of selectors, which are already known by the
compiler's template builder, and passes them through to calls to
`defineComponent()` so that they are available on the component
factory.

## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?
<!-- Please describe the current behavior that you are modifying, or link to a relevant issue. -->

Issue Number: N/A


## What is the new behavior?


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

petebacondarwin commented:
Currently there are now  4 failing tests here - but I think they are because of other issues.

mary-poppins commented:
You can preview 9aed893 at https://pr27769-9aed893.ngbuilds.io/.

mary-poppins commented:
You can preview 98babc1 at https://pr27769-98babc1.ngbuilds.io/.

AndrewKushnir commented:
Hi @petebacondarwin, the logic that passed `ngContentSelectors` seems right, however it'd be great to rebase this PR once Pawel's PR is merged: https://github.com/angular/angular/pull/27755. That PR brings all the necessary content selectors to the top level template, so you'll be able to just expose it via a function (similar to what you have now).

mary-poppins commented:
You can preview 0fcc19b at https://pr27769-0fcc19b.ngbuilds.io/.

pkozlowski-opensource commented on packages/upgrade/test/common/test_helpers.ts:
```diff
@@ -13,14 +13,14 @@ const TEST_MINIFIED = true;
-  {
-    label: '1.6',
-    files: [`angular-1.6/${ANGULARJS_FILENAME}`, 'angular-mocks-1.6/angular-mocks.js'],
-  },
```
Was it commented out intentionally?

petebacondarwin replied:
Yes, this is marked WIP :-P

pkozlowski-opensource commented on packages/compiler/src/render3/view/compiler.ts:
```diff
@@ -260,6 +260,13 @@ export function compileComponentFromMetadata(
+  const ngContentSelectors = templateBuilder.getNgContentSelectors();
+  if (ngContentSelectors) {
+    definitionMap.set('ngContentSelectors', ngContentSelectors);
+  }
```
Could you please add a compiler-specific test(s) that asserts on the generated code? I would like to see what is expected generated but I suspect that we are going to duplicate CSS selectors list(s) - once here and second time in the generated code, right? If so, would be great to see if we can avoid it..

Other question I'm having hard time answering without a test - are putting both parsed and un-parsed selectors in `ngContentSelectors`? If so, do you need both forms on the upgrade side?

petebacondarwin replied:
Yes, definitely.

pkozlowski-opensource reviewed (changes requested):
Agree with @AndrewKushnir that it should be rebased on top of #27755 as we are changing algorighm for collecting selectors in there.

I would also love to see a compiler-specific test to see what code is generated, exactly - left more details in the code comments

mary-poppins commented:
You can preview 9e31f19 at https://pr27769-9e31f19.ngbuilds.io/.

petebacondarwin committed:
fix(ivy): pass ngContentSelectors through to `defineComponent()` calls

Libraries like `@angular/upgrade` need to pass blocks of projected content
through to components that they dynamically instantiate. These blocks
are selected by CSS selectors defined in `<ng-content select="..">`
tags found in the components template.

The Angular compiler collects these CSS selectors when compiling a component's
template, but we need to then expose these on the compiled component factory
to make them available to libraries such as `@angular/upgrade`.

This change takes the list of selectors, which were found by the
compiler's template builder, and passes them through to calls to
`defineComponent()`, onto the component definition, and finally to the
component factory.

petebacondarwin committed:
fix(ivy): component factory should pass correct projectable nodes

`ComponentFactory.create()` can receive an array of projectable nodes
as one of its arguments. This array was not being mapped correctly to the
buckets that are expected by the `projectionDef` and `project` ivy
instructions.

This commit re-organizes the `projectableNodes` array to match what is
expected by ivy.

Fixes #27791
Fixes FW-886

mary-poppins commented:
You can preview 3c1c79e at https://pr27769-3c1c79e.ngbuilds.io/.

mary-poppins commented:
You can preview cf06ac0 at https://pr27769-cf06ac0.ngbuilds.io/.

mary-poppins commented:
You can preview 149d47e at https://pr27769-149d47e.ngbuilds.io/.

petebacondarwin commented:
Closing as the problem with ngContentSelectors is multi-layered and this PR is trying to solve too much. Instead, I will open a simple PR that fixes the problem for ngUpgrade use-cases. Then we can discuss the fact that the current ordering of ngContentSelectors, in Ivy mode, is different to that in ViewEngine mode.
